contrib/revoke_legacy.sh

```diff
@@ -0,0 +1,105 @@
+fi
+
+# dcrdata_db_name is the name of the dcrdata database.
+dcrdata_db_name="dcrdata_${network}"
+
+# psql is the base psql command to run including the database user, database
+# name, and flags.
+psql="psql -U dcrdata -d ${dcrdata_db_name} -XAtc"
+
+# unrevoked_sql_count is the query for fetching the total unrevoked tickets
+# count.
+unrevoked_sql_count="SELECT count(*) "
+unrevoked_sql_count+="FROM misses "
+unrevoked_sql_count+="INNER JOIN tickets ON ticket_hash = tx_hash "
+unrevoked_sql_count+="WHERE spend_tx_db_id IS NULL"
+
+# unrevoked_sql is the query for fetching the next batch of unrevoked tickets.
+unrevoked_sql="SELECT ticket_hash, price "
+unrevoked_sql+="FROM misses "
+unrevoked_sql+="INNER JOIN tickets ON ticket_hash = tx_hash "
+unrevoked_sql+="WHERE spend_tx_db_id IS NULL "
+unrevoked_sql+="ORDER BY height, ticket_hash "
+unrevoked_sql+="LIMIT ${batch_size}"
+
+# Create ticket revocation transactions in batches per-block until there are no
+# unrevoked tickets remaining.
+while :
+do
+  # Fetch the best block height.  Continue if the block height has not changed
+  # so that only one batch of revocation transactions are broadcast per block.
+  height=$(${psql} "SELECT best_block_height FROM meta")
+  if [ -n "$prev_height" ] && [ $height == $prev_height ]; then
+    # Sleep in order to only query for the best block height every 5 seconds.
+    sleep 5
+    continue
+  fi
+
+  # Fetch the unrevoked tickets count.  Exit if it is zero since there is
+  # nothing left to do.
+  unrevoked_count=($(${psql} "${unrevoked_sql_count}"))
+  if [ $unrevoked_count == 0 ]; then
+    echo "No remaining unrevoked tickets... Done!"
+    exit 0
+  fi
+
+  # Process the next batch of unrevoked tickets.  The unrevoked tickets are
+  # queried each block (rather than fetching them once up front) since
+  # revocations that aren't included in a block will be dropped from the
+  # mempool.
+  unrevoked_misses=($(${psql} "${unrevoked_sql}"))
+  status_msg="Processing the next ${#unrevoked_misses[@]} unrevoked tickets... "
+  status_msg+="(Best Height: ${height}, Total Unrevoked: ${unrevoked_count})"
+  echo $status_msg
+  for unrevoked in "${unrevoked_misses[@]}"
+  do
+    # Get the ticket hash and price from the query result.
+    IFS='|' read -ra unrevoked_arr <<< "$unrevoked"
+    ticket_hash=${unrevoked_arr[0]}
+    price=${unrevoked_arr[1]}
+
+    # Construct the createrawssrtx command.
+    json="[{\"amount\":$price,\"txid\":\"${ticket_hash}\",\"vout\":0,\"tree\":1}]"
+    create_raw_ssrtx_command="${dcrctl_cmd} createrawssrtx ${json}"
+    echo $create_raw_ssrtx_command
+
+    # Run the createrawssrtx command.
+    revocation_tx_hex=($($create_raw_ssrtx_command))
+    echo $revocation_tx_hex
+
+    # Construct the sendrawtransaction command.
+    send_raw_tx_command="${dcrctl_cmd} sendrawtransaction ${revocation_tx_hex}"
+
+    # Run the sendrawtransaction command.
+    revocation_tx_hash=($($send_raw_tx_command))
```

Comment:
Probably should check for an error here.  Since it's sitting in a loop, if there is a reorg or a block is mined without the transaction (there is nothing forcing the legacy revocations to be included), you'll definitely end up with some attempts at creating duplicates at a minimum.

I suppose since this is just mostly a manual operation allowing it to fail and just logging the error is probably fine too.

Reply:
Part of the reason why I have the script re-querying the unrevoked tickets at each block interval is that at each block, the revocations should either be mined or [dropped from the mempool](https://github.com/decred/dcrd/blob/64aabc0283e9c9d2b0e548b754417b66e3c23629/internal/mempool/mempool.go#L2063-L2072) (so this should avoid submitting duplicates).

Currently, if there is an error, it will just print to stdout and continue as you mentioned.  Then at the next block interval, it will either try again (if a revocation for that ticket wasn't mined), or it will have already been mined and therefore won't be returned again in the query.  I thought this was probably best because it should eventually succeed unless there is a bug in the script, but I'm open to suggestions of other ways to handle it.

Reply:
Ah, good point regarding the mempool. I forgot about the new requirement on the previous header, so it will indeed be dropped.

I believe this should work fine.  We can always just manually deal (or update the script accordingly) with anything comes up as well of course.

Reply:
Yep, sounds good.  Thanks for taking a look!